torchao/prototype/optim_8bit/adam.py

```diff
@@ -0,0 +1,200 @@
+        amsgrad=False,
+        block_size=2048,
+    ):
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))
+        defaults = dict(lr=lr, betas=betas, eps=eps, weight_decay=weight_decay, amsgrad=amsgrad)
+        super().__init__(params, defaults)
+        self.block_size = block_size
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("amsgrad", False)
+
+    # follow bitsandbytes
+    # only apply quantization for tensor with more than 4096 values
+    # TODO: also skip 1D tensor? e.g. biases and norm scales
+    def _new_zero_buffer(self, p: Tensor, signed: bool = True):
+        if p.numel() >= 4096 and p.numel() % self.block_size == 0:
+            out = DTQ8bit.zeros(p.shape, signed, self.block_size, device=p.device)
+        else:
+            out = torch.zeros_like(p)
+        return out
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Sparse gradient is not supported")
+
+                state = self.state[p]
+
+                # State initialization
+                # state is flattened so that torch.compile won't recompile for tensors with different ndim
+                if len(state) == 0:
+                    state["step"] = torch.tensor(0.0, device=p.device)
+                    state["exp_avg"] = self._new_zero_buffer(p.view(-1), signed=True)
+                    state["exp_avg_sq"] = self._new_zero_buffer(p.view(-1), signed=False)
+                    if group["amsgrad"]:
+                        state["max_exp_avg_sq"] = self._new_zero_buffer(p.view(-1), signed=False)
+
+                state["step"] += 1
+
+                # flatten p and grad so that torch.compile won't recompile for tensors with different ndim
+                # must explicitly convert lr to Tensor since torch.compile() will treat it as a constant
+                # if it is a python float. practically, only lr is changed during training.
+                # NOTE: if lr is change at every step, moving lr to CUDA will be a bottleneck.
+                if not isinstance(group["lr"], Tensor):
+                    group["lr"] = torch.tensor(group["lr"], device=p.device)
```

Comment:
If I understand this code correctly, the code right now initiates lr as a Tensor on the device of the current param just once per group, yes?

This assumes that all params in one param group have the same device, yes?

Reply:
> If I understand this code correctly, the code right now initiates lr as a Tensor on the device of the current param just once per group, yes? And this fixes the recompile problem when lr changes value?

Yes.

> This assumes that all params in one param group have the same device, yes?

Yes. I hope this is a reasonable assumption. In what scenario this assumption would not be valid? e.g. FSDP?